runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/smithy/Document.kt

```diff
@@ -4,161 +4,185 @@
+    }
 
     /**
-     * Returns the content as a double which may involve rounding
+     * Wraps a [kotlin.collections.Map].
      */
-    val double: Double get() = content.toDouble()
-
-    override fun toString(): String = content.toString()
-}
+    data class Map(val value: kotlin.collections.Map<kotlin.String, Document>) : Document() {
+        override fun toString() = value
+            .entries
+            .joinToString(
+                separator = ",",
+                prefix = "{",
+                postfix = "}",
+                transform = { (k, v) -> """"$k":$v""" }
+            )
+    }
 
-/**
- * Class representing document `array` type
- */
-data class SmithyArray(val content: List<Document>) : Document(), List<Document> by content {
     /**
-     * Returns [index] th element of an array as [SmithyNumber] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
+     * Represents a `null` value.
      */
-    fun getNumber(index: Int) = content[index] as? SmithyNumber
+    object Null : Document() {
+        override fun toString() = "null"
+    }
 
-    /**
-     * Returns [index] th element of an array as [SmithyBool] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
-     */
-    fun getBoolean(index: Int) = content[index] as? SmithyBool
+    fun asNumber() = (this as Number).value
+    fun asString() = (this as String).value
+    fun asBoolean() = (this as Boolean).value
+    fun asList() = (this as List).value
+    fun asMap() = (this as Map).value
 
-    /**
-     * Returns [index] th element of an array as [SmithyString] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
-     */
-    fun getString(index: Int) = content[index] as? SmithyString
+    fun asInt() = asNumber().toInt()
+    fun asByte() = asNumber().toByte()
+    fun asLong() = asNumber().toLong()
+    fun asFloat() = asNumber().toFloat()
+    fun asDouble() = asNumber().toDouble()
 
-    /**
-     * Returns [index] th element of an array as [SmithyArray] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
-     */
-    fun getArray(index: Int) = content[index] as? SmithyArray
+    val isNull: kotlin.Boolean
+        get() = this == Null
+
+    operator fun get(i: Int) = asList()[i]
+    operator fun get(i: kotlin.String) = asMap()[i]
 
     /**
-     * Returns [index] th element of an array as [SmithyMap] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
+     * DSL builder for a [Map]-based [Document].
      */
-    fun getMap(index: Int) = content[index] as? SmithyMap
+    class Builder internal constructor() {
```

Comment:
Not a hill I'm going to die on but I think I liked the document builders separate (especially since we've moved the definition of each variant into the sealed class body making it harder to read)

Reply:
This was a question of consistency for me - all of our request objects localize the builder within the class itself.

Reply:
Our generated code doesn't have the luxury of being hand written though. The choices we make for generated code deal with a lot of other concerns. 

Like I said this isn't a huge deal so if you feel strongly I'm fine with it here.